README.md
[![build status](https://secure.travis-ci.org/openmason/jsonrpclib.png)](http://travis-ci.org/openmason/jsonrpclib)
# json rpc library
json rpc 2.0 server library.
For a ready to use ZeroMQ-JSON-RPC server, please refer to https://github.com/openmason/zmqrpc

# Features
 * full implementation of json rpc 2.0 spec (see the spec http://www.jsonrpc.org/specification for details)
 * support for both positional and named arguments
 * support for batch request/response

# Usage
Please refer to test directory for full suite of test cases and how to use.

    # define a simple module
    var MyService = {
      add:      function(a,b) { return a+b; },
      subtract: function(a,b) { return a-b; }
    };
    
    var jsonrpclib=require('jsonrpclib');
    var rpc = new jsonrpclib(MyService);
    
    // call the service
    res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}');
    // res is a json string and would be like
    // {"jsonrpc":"2.0","result":19,"id":"1"}

# Tests
Full set of test cases can be found in test directory. To run tests just issue `make test`
    


History.md
# release 0.1.3
 * fixed issue during empty params
 * full implementaion of rpc 2.0 spec 
 * unit test cases added

# release 0.0.1
 * initial repository creation


test/basic.js
var jsonrpclib=require('..');
var assert=require('assert');

// nothing for now
/*
 * Request
   * jsonrpc - "2.0"
   * method 
 * Response
 * Batch
*/
describe('basic', function() {
  before(function(done) {
    done();
  });
  var testModules = {
    add: function(a,b) { return a+b; },
    subtract: function(a,b) { return a-b; },
    name: 'test module'
  };
  var rpc=new jsonrpclib(testModules);
  rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1');
  rpc.request('[]');
  rpc.request('{}');
  rpc.request('{id:1}');
  rpc.request('{"jsonrpc":"1.2"}');
  rpc.request('{"jsonrpc":"1.2", "id":1}');
  rpc.request('{"jsonrpc":"2.0"}');
  rpc.request('{"jsonrpc":"2.0", "id":1}');
  rpc.request('{"jsonrpc":"2.0", "id":[1]}');
  rpc.request('{"jsonrpc":"2.0", "params":[1,2], "id":123}'); // method missing
  rpc.request('{"jsonrpc":"2.0", "mehod":"subtract", "params":[1,2], "id":1}'); // method missing
  rpc.request('{"jsonrpc":"2.0", "method":"sub", "params":[1,2], "id":1}'); // method unknown
  rpc.request('{"jsonrpc":"2.0", "method":"name", "params":[1,2], "id":1}'); // not a method
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":23, "id":1}'); // invalid param
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":"2,3", "id":1}'); // invalid param
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[5,2], "id":1}');
  // params should match
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[5,2,3], "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[], "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[5], "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"a":10}, "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"a":10,"x":3}, "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"a":10,"b":3,"c":34}, "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"b":3,"a":34}, "id":1}');
  rpc.request('{"jsonrpc":"2.0", "method":"add", "params":{"b":3,"a":34}, "id":1}');
  
  //
  rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}');
  // batch tests
  var r=rpc.request('[{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1},{"sample":"value"}]');
  r=rpc.request('[{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1},{"method":"subtract","id":23}]');
  console.log(r);

  // lets copy & paste all the examples
  testModules = {
    sum: function(a,b,c) { return a+b+c; },
    subtract: function(minuend,subtrahend) { return minuend-subtrahend; },
    get_data: function() { return ['hello',5]; }
  };
  rpc=new jsonrpclib(testModules);

  // from the spec here
  // rpc call with positional parameters
  r=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}');
  console.log(r);
  r=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [23, 42], "id": 2}');
  console.log(r);

  // rpc call with named parameters
  r=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 3}');
  console.log(r);
  r=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": {"minuend": 42, "subtrahend": 23}, "id": 4}');
  console.log(r);

  // a Notification:
  r=rpc.request('{"jsonrpc": "2.0", "method": "update", "params": [1,2,3,4,5]}');
  console.log(r);
  r=rpc.request('{"jsonrpc": "2.0", "method": "foobar"}');
  console.log(r);

  // rpc call of non-existent method:
  r=rpc.request('{"jsonrpc": "2.0", "method": "foobar", "id": "1"}');
  console.log(r);

  // rpc call with invalid JSON:
  r=rpc.request('{"jsonrpc": "2.0", "method": "foobar, "params": "bar", "baz]');
  console.log(r);

  // rpc call with invalid Request object:
  r=rpc.request('{"jsonrpc": "2.0", "method": 1, "params": "bar"}');
  console.log(r);

  // rpc call Batch, invalid JSON:
  r=rpc.request('[ {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},{"jsonrpc": "2.0", "method"');
  console.log(r);

  // rpc call with an invalid Batch (but not empty):
  r=rpc.request('[1]');
  console.log(r);

  // rpc call with invalid Batch:
  r=rpc.request('[1,2,3]');
  console.log(r);

  // rpc call Batch:
  r=rpc.request('[{"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},'
                + '{"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},'
                + '{"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},'
                + '{"foo": "boo"},'
                + '{"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"}, '
                + '{"jsonrpc": "2.0", "method": "get_data", "id": "9"} '
                + ']');
  console.log(r);

  // rpc call Batch (all notifications):
  r=rpc.request('[ '
                + '{"jsonrpc": "2.0", "method": "notify_sum", "params": [1,2,4]}, '
                + '{"jsonrpc": "2.0", "method": "notify_hello", "params": [7]} '
                + ']');
  console.log(r);

});


test/rpclib.js
var jsonrpclib=require('..');
var assert=require('assert');

function _getCodeOrResult(jsonStr, code) {
  if(!jsonStr) return jsonStr;
  try {
    var obj = JSON.parse(jsonStr);
    return obj.error ? obj.error.code : obj.result;
  }
  catch(err) {
    // json has to be valid in all cases
    throw Error('Invalid JSON');
  }
}

var PARSE_ERROR = -32700;
var INVALID_REQUEST = -32600;
var METHOD_NOT_FOUND = -32601;
var INVALID_PARAMS = -32602;


describe('json-rpc library', function() {
  before(function(done) {
    done();
  });
  var testModules = {
    add: function(a,b) { return a+b; },
    subtract: function(a,b) { return a-b; },
    getName: function() { return 'hi'; },
    name: 'test module'
  };
  var debug=false;
  describe('basic variants', function() {
    it('syntax', function(done) {
      var rpc=new jsonrpclib(testModules,debug);
      var res;
      res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1');
      assert.equal(_getCodeOrResult(res), PARSE_ERROR);

      res=rpc.request('[]');
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{}');
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{"id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{"jsonrpc":"1.2"}');
      assert.equal(_getCodeOrResult(res), undefined);

      res=rpc.request('{"jsonrpc":"1.2", "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{"jsonrpc":"2.0"}');
      assert.equal(_getCodeOrResult(res), undefined);

      res=rpc.request('{"jsonrpc":"2.0", "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{"jsonrpc":"2.0", "id":[1]}');
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      done();

    });
    it('method', function(done) {
      var rpc=new jsonrpclib(testModules,debug);
      var res;
      res=rpc.request('{"jsonrpc":"2.0", "params":[1,2], "id":123}'); // method missing
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{"jsonrpc":"2.0", "mehod":"subtract", "params":[1,2], "id":1}'); // method missing
      assert.equal(_getCodeOrResult(res), INVALID_REQUEST);

      res=rpc.request('{"jsonrpc":"2.0", "method":"sub", "params":[1,2], "id":1}'); // method unknown
      assert.equal(_getCodeOrResult(res), METHOD_NOT_FOUND);

      res=rpc.request('{"jsonrpc":"2.0", "method":"name", "params":[1,2], "id":1}'); // not a method
      assert.equal(_getCodeOrResult(res), METHOD_NOT_FOUND);
      done();
    });
    it('params', function(done) {
      var rpc=new jsonrpclib(testModules,debug);
      var res;
      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":23, "id":1}'); // invalid param
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":"2,3", "id":1}'); // invalid param
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[5,2], "id":1}');
      assert.equal(_getCodeOrResult(res), 3);

      // params should match
      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[5,2,3], "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[], "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":[5], "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"a":10}, "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"a":10,"x":3}, "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"a":10,"b":3,"c":34}, "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"subtract", "params":{"b":3,"a":34}, "id":1}');
      assert.equal(_getCodeOrResult(res), 31);

      res=rpc.request('{"jsonrpc":"2.0", "method":"add", "params":{"b":3,"a":34}, "id":1}');
      assert.equal(_getCodeOrResult(res), 37);

      res=rpc.request('{"jsonrpc":"2.0", "method":"getName", "id":1}');
      assert.equal(_getCodeOrResult(res), 'hi');

      res=rpc.request('{"jsonrpc":"2.0", "method":"getName", "params":{}, "id":1}');
      assert.equal(_getCodeOrResult(res), 'hi');

      res=rpc.request('{"jsonrpc":"2.0", "method":"getName", "params":[], "id":1}');
      assert.equal(_getCodeOrResult(res), 'hi');

      res=rpc.request('{"jsonrpc":"2.0", "method":"getName", "params":[1], "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      res=rpc.request('{"jsonrpc":"2.0", "method":"getName", "params":{"a":12}, "id":1}');
      assert.equal(_getCodeOrResult(res), INVALID_PARAMS);

      // valid
      res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}');
      assert.equal(_getCodeOrResult(res), 19);

      done();
    });
    // batch tests
    it('batch', function(done) {
      var rpc=new jsonrpclib(testModules,debug);
      var res;
      res=rpc.request('[{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1},{"sample":"value"}]');
      var obj=JSON.parse(res);
      assert.equal(obj.length, 1);
      assert.equal(obj[0].result, 19);

      res=rpc.request('[{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1},{"method":"subtract","id":23}]');
      obj=JSON.parse(res);
      assert.equal(obj.length, 2);
      assert.equal(obj[0].result, 19);
      assert.equal(obj[1].error.code, INVALID_PARAMS);
      done();
    });

  });

  // lets copy & paste all the examples from rpc spec page
  var testModules1 = {
    sum: function(a,b,c) { return a+b+c; },
    subtract: function(minuend,subtrahend) { return minuend-subtrahend; },
    get_data: function() { return ['hello',5]; }
  };
  describe('rpc spec examples', function() {
    var rpc=new jsonrpclib(testModules1,debug);
    var res;
    it('rpc call with positional parameters', function(done) {
      res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}');
      assert.equal(_getCodeOrResult(res), 19);
      res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": [23, 42], "id": 2}');
      assert.equal(_getCodeOrResult(res), -19);
      done();
    });
    it('rpc call with named parameters', function(done) {
      res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 3}');
      assert.equal(_getCodeOrResult(res), 19);
      res=rpc.request('{"jsonrpc": "2.0", "method": "subtract", "params": {"minuend": 42, "subtrahend": 23}, "id": 4}');
      assert.equal(_getCodeOrResult(res), 19);
      done();
    });
    it('a Notification', function(done) {
      res=rpc.request('{"jsonrpc": "2.0", "method": "update", "params": [1,2,3,4,5]}');
      assert.equal(_getCodeOrResult(res), undefined);
      res=rpc.request('{"jsonrpc": "2.0", "method": "foobar"}');
      assert.equal(_getCodeOrResult(res), undefined);
      done();
    });
    it('rpc call of non-existent method', function(done) {
      res=rpc.request('{"jsonrpc": "2.0", "method": "foobar", "id":1}');
      assert.equal(_getCodeOrResult(res), METHOD_NOT_FOUND);
      done();
    });
    it('rpc call with invalid JSON', function(done) {
      res=rpc.request('{"jsonrpc": "2.0", "method": "foobar, "params": "bar", "baz]');
      assert.equal(_getCodeOrResult(res), PARSE_ERROR);
      done();
    });
    it('rpc call with invalid Request object', function(done) {
      res=rpc.request('{"jsonrpc": "2.0", "method": 1, "params": "bar"}');
      assert.equal(_getCodeOrResult(res), undefined);
      done();
    });
    it('rpc call Batch, invalid JSON', function(done) {
      res=rpc.request('[ {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},{"jsonrpc": "2.0", "method"');
      assert.equal(_getCodeOrResult(res), PARSE_ERROR);
      done();
    });
    it('rpc call with an invalid Batch (but not empty)', function(done) {
      res=rpc.request('[1]');
      var obj=JSON.parse(res);
      assert.equal(obj.length, 1);
      assert.equal(obj[0].error.code, INVALID_REQUEST);
      done();
    });
    it('rpc call with invalid Batch', function(done) {
      res=rpc.request('[1,2,3]');
      var obj=JSON.parse(res);
      assert.equal(obj.length, 3);
      assert.equal(obj[0].error.code, INVALID_REQUEST);
      assert.equal(obj[1].error.code, INVALID_REQUEST);
      assert.equal(obj[2].error.code, INVALID_REQUEST);
      done();
    });
    it('rpc call Batch', function(done) {
      res=rpc.request('[{"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},'
                      + '{"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},'
                      + '{"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},'
                      + '{"foo": "boo"},'
                      + '{"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"}, '
                      + '{"jsonrpc": "2.0", "method": "get_data", "id": "9"} '
                      + ']');
      var obj=JSON.parse(res);
      assert.equal(obj.length, 4);
      // order might not be preserved in actuality
      assert.equal(obj[0].result,7);
      assert.equal(obj[1].result,19);
      assert.equal(obj[2].error.code, METHOD_NOT_FOUND);
      assert.equal(obj[3].id, "9");
      done();
    });
    it('rpc call Batch (all notifications)', function(done) {
      res=rpc.request('[ '
                    + '{"jsonrpc": "2.0", "method": "notify_sum", "params": [1,2,4]}, '
                    + '{"jsonrpc": "2.0", "method": "notify_hello", "params": [7]} '
                    + ']');
      assert.equal(_getCodeOrResult(res), undefined);
      done();
    });

  });


});


index.js
/**
 * jsonrpclib - main file
 * copyright (c) 2012 openmason.
 * MIT Licensed.
 */

var handy = require('handy');
var logger = require('util');
var _ = require('underscore');

// The rpc specification can be found at:
// - http://www.jsonrpc.org/specification
// - http://en.wikipedia.org/wiki/JSON-RPC

/*
 * Syntax:
 * 
 * --> data sent to Server
 * <-- data sent to Client
 * 
 * rpc call with positional parameters:
 * 
 * --> {"jsonrpc": "2.0", "method": "subtract", "params": [42, 23], "id": 1}
 * <-- {"jsonrpc": "2.0", "result": 19, "id": 1}
 * 
 * --> {"jsonrpc": "2.0", "method": "subtract", "params": [23, 42], "id": 2}
 * <-- {"jsonrpc": "2.0", "result": -19, "id": 2}
 * 
 * rpc call with named parameters:
 * 
 * --> {"jsonrpc": "2.0", "method": "subtract", "params": {"subtrahend": 23, "minuend": 42}, "id": 3}
 * <-- {"jsonrpc": "2.0", "result": 19, "id": 3}
 * 
 * --> {"jsonrpc": "2.0", "method": "subtract", "params": {"minuend": 42, "subtrahend": 23}, "id": 4}
 * <-- {"jsonrpc": "2.0", "result": 19, "id": 4}
 * 
 * a Notification:
 * 
 * --> {"jsonrpc": "2.0", "method": "update", "params": [1,2,3,4,5]}
 * --> {"jsonrpc": "2.0", "method": "foobar"}
 * 
 * rpc call of non-existent method:
 * 
 * --> {"jsonrpc": "2.0", "method": "foobar", "id": "1"}
 * <-- {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found."}, "id": "1"}
 * 
 * rpc call with invalid JSON:
 * 
 * --> {"jsonrpc": "2.0", "method": "foobar, "params": "bar", "baz]
 * <-- {"jsonrpc": "2.0", "error": {"code": -32700, "message": "Parse error."}, "id": null}
 * 
 * rpc call with invalid Request object:
 * 
 * --> {"jsonrpc": "2.0", "method": 1, "params": "bar"}
 * <-- {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null}
 * 
 * rpc call Batch, invalid JSON:
 * 
 * --> [
 *   {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
 *   {"jsonrpc": "2.0", "method"
 * ]
 * <-- {"jsonrpc": "2.0", "error": {"code": -32700, "message": "Parse error."}, "id": null}
 * rpc call with an empty Array:
 * --> []
 * <-- {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null}
 * rpc call with an invalid Batch (but not empty):
 * --> [1]
 * <-- [
 *   {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null}
 * ]
 * 
 * rpc call with invalid Batch:
 * 
 * --> [1,2,3]
 * <-- [
 *   {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null},
 *   {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null},
 *   {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null}
 * ]
 * 
 * rpc call Batch:
 * 
 * --> [
 *         {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
 *         {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
 *         {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
 *         {"foo": "boo"},
 *         {"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
 *         {"jsonrpc": "2.0", "method": "get_data", "id": "9"} 
 *     ]
 * <-- [
 *         {"jsonrpc": "2.0", "result": 7, "id": "1"},
 *         {"jsonrpc": "2.0", "result": 19, "id": "2"},
 *         {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request."}, "id": null},
 *         {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found."}, "id": "5"},
 *         {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
 *     ]
 * 
 * rpc call Batch (all notifications):
 * 
 * --> [
 *         {"jsonrpc": "2.0", "method": "notify_sum", "params": [1,2,4]},
 *         {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]}
 *     ]
 * <-- //Nothing is returned for all notification batches
 * 
 */

var RpcErrors = {
  PARSE_ERROR     : { code:-32700, message: 'Parse error' },
  INVALID_REQUEST : { code:-32600, message: 'Invalid Request' },
  METHOD_NOT_FOUND: { code:-32601, message: 'Method not found' },
  INVALID_PARAMS  : { code:-32602, message: 'Invalid params' },
  INTERNAL_ERROR  : { code:-32603, message: 'Internal error' },
  SERVER_ERROR    : { code:-32000, message: 'Server error' }
};

// JSON RPC library entry point
var JSONRpc = function(module, debug) {
  this.debug = debug || false;
  this.jsonrpc = "2.0";
  // check & load the methods in module
  this.methods = module;
  if(handy.getType(module)=='string') {
    this.methods = require(module);
  }
  if(this.debug) {
    logger.debug('Loaded with methods:'+_.functions(this.methods));
  }
};

// main dispatcher for processing json-rpc
// requests
JSONRpc.prototype.request = function(jsonBody) {
  var self=this;
  self._debug(true, jsonBody);

  var id = null;
  var batch = false;
  var rpcObj;
  // first step is to parse the json
  try {
    rpcObj = JSON.parse(jsonBody);
  } catch(err) {
    return JSON.stringify(self.error(RpcErrors.PARSE_ERROR, id, err.message));
  }
  var requests = [];
  var results = [];
  // if rpcObj is array, then its a batch request
  if(handy.getType(rpcObj)=='array') {
    if(rpcObj.length==0) {
      return JSON.stringify(self.error(RpcErrors.INVALID_REQUEST, id));
    }
    batch = true;
    requests = rpcObj;
  } else {
    requests = [rpcObj];
  }
  // handle all the requests
  // 1. try to validate each of the request
  // 2. run them
  _.each(requests, function(reqObj) {
    // - first drop off all the requests without id
    //   those are notifications.
    if(handy.getType(reqObj)=='object' && _.size(reqObj)>0 && !_.has(reqObj, 'id')) {
      self._debug(true, 'Notification ' + JSON.stringify(reqObj));
    } else {
      var result = self._validate(reqObj);
      if(!result) {
        // invoke the function
        // @todo - what should be the value of 'this'?
        try {
          var res=self.methods[reqObj.method].apply(null, reqObj.params);
          result = self.result(reqObj.id, res);
        }
        catch(err) {
          result = self.error(RpcErrors.INTERNAL_ERROR, reqObj.id, err);
        }
      }
      results.push(result);
    }
  });
  // return back the result
  if(results.length<=0) return;
  if(batch==false) {
    return JSON.stringify(results[0]);
  } 
  return JSON.stringify(results);
};

// return back the result object
JSONRpc.prototype.result = function(id, result) {
  var res = { jsonrpc: this.jsonrpc,
              result: result,
              id: id };
  this._debug(false, res);
  return res;
};


// return back the correct error object
JSONRpc.prototype.error = function(err, id, data) {
  var errorObj = { jsonrpc: this.jsonrpc,
                   error: { code: err.code, message: err.message },
                   id: id };
  if(data) {
    errorObj['data'] = data;
  }
  this._debug(false, errorObj);
  return errorObj;
};

// ---- private functions

// validate the request object
// - returns the error object, if any error
JSONRpc.prototype._validate = function(requestObj) {
  var self = this;
  // - check for jsonprc value
  if(_.has(requestObj, 'jsonrpc') && requestObj['jsonrpc']!='2.0') {
    return self.error(RpcErrors.INVALID_REQUEST, requestObj.id, 'unknown jsonrpc version');
  }
  // - check for id
  var idType = handy.getType(requestObj.id);
  if(idType != 'string' && idType != 'number') {
    return self.error(RpcErrors.INVALID_REQUEST, requestObj.id, 'id should be a valid number/string');
  }
  // - check for method
  if(!_.has(requestObj, 'method')) {
    return self.error(RpcErrors.INVALID_REQUEST, requestObj.id, 'missing method to call');
  }
  // - check if method is present
  var fns = _.functions(self.methods);
  if(!_.include(fns, requestObj.method)) {
    return self.error(RpcErrors.METHOD_NOT_FOUND, requestObj.id, requestObj.method + " - unknown method");
  }

  // - parameter checks
  var params=_getParamNames(self.methods[requestObj.method]) || [];
  // - if params are absent and required for the method, its an error
  if(!_.has(requestObj,'params') && params && params.length>0) {
    return self.error(RpcErrors.INVALID_PARAMS, requestObj.id, 'params expected:'+params);
  }
  // - if params are present
  //   it has to be either array or object
  if(_.has(requestObj, 'params')) {
    var ptype = handy.getType(requestObj.params);
    if(ptype!='array' && ptype!='object') {
      return self.error(RpcErrors.INVALID_PARAMS, requestObj.id, 'params should be either array or object');
    }

    // @todo - not sure if this check needs to be enabled
    // sometimes it might be by design that less valus can be passed

    // check if array matches the arguments
    if(ptype=='array' && requestObj.params.length != params.length) {
      return self.error(RpcErrors.INVALID_PARAMS, requestObj.id, 'total params expected:'+params.length);
    }
    // check if the object has matching params
    if(ptype=='object') {
      var requestValues = _.keys(requestObj.params);
      if(!handy.isArrayEqual(params, requestValues)) {
        return self.error(RpcErrors.INVALID_PARAMS, requestObj.id, 'params expected:'+params);
      }
      // lets convert the params to array 
      // in the order expected
      requestObj.params = _.values(_.pick(requestObj.params, params));
    }
  }
};

// returns the function parameters
function _getParamNames(func) {
  var funStr = func.toString();
  return funStr.slice(funStr.indexOf('(')+1, funStr.indexOf(')')).match(/([^\s,]+)/g);
}

// debug request/response statements
JSONRpc.prototype._debug = function(isRequest, value) {
  if(this.debug) {
    if(handy.getType(value)!='string') {
      value = JSON.stringify(value);
    }
    logger.debug((isRequest?'-->':'<--')+' ' + value);
  }
};

module.exports = JSONRpc;

// -- EOF